File: python/tree-building/tree_building.py
class Record:
    def __init__(self, record_id, parent_id):
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    def __init__(self, node_id):
        self.node_id = node_id
        self.children = []


def BuildTree(records):
    # Handle empty input
    if not records:
        return None
    
    # Sort records by ID for easier processing
    records.sort(key=lambda x: x.record_id)
    
    # Validate records
    if records[0].record_id != 0:
        raise ValueError("Record id is invalid or out of order.")
    
    if records[-1].record_id != len(records) - 1:
        raise ValueError("Record id is invalid or out of order.")
    
    # Create nodes for each record
    nodes = {}
    for record in records:
        # Validate record
        if record.record_id != record.parent_id and record.record_id < record.parent_id:
            raise ValueError("Node parent_id should be smaller than it's record_id.")
        
        if record.record_id == record.parent_id and record.record_id != 0:
            raise ValueError("Only root should have equal record and parent id.")
        
        # Create node
        nodes[record.record_id] = Node(record.record_id)
    
    # Build tree by connecting nodes
    for record in records:
        # Skip root node (it has no parent)
        if record.record_id == 0:
            continue
        
        # Connect node to its parent
        parent = nodes[record.parent_id]
        parent.children.append(nodes[record.record_id])
    
    return nodes[0] if nodes else None

File: go/tree-building/tree_building.go
package tree

import (
	"errors"
	"sort"
)

type Record struct {
	ID     int
	Parent int
}

type Node struct {
	ID       int
	Children []*Node
}

func Build(records []Record) (*Node, error) {
	if len(records) == 0 {
		return nil, nil
	}

	// Sort records by ID for easier validation and building
	sort.Slice(records, func(i, j int) bool {
		return records[i].ID < records[j].ID
	})

	// Validate records
	if err := validateRecords(records); err != nil {
		return nil, err
	}

	// Create nodes map
	nodes := make(map[int]*Node, len(records))
	for _, record := range records {
		nodes[record.ID] = &Node{ID: record.ID}
	}

	// Build the tree by connecting nodes
	for _, record := range records {
		// Skip the root node (ID 0, Parent 0)
		if record.ID == 0 {
			continue
		}

		parent := nodes[record.Parent]
		parent.Children = append(parent.Children, nodes[record.ID])
	}

	// Sort children by ID for each node
	for _, node := range nodes {
		if len(node.Children) > 0 {
			sort.Slice(node.Children, func(i, j int) bool {
				return node.Children[i].ID < node.Children[j].ID
			})
		}
	}

	return nodes[0], nil
}

func validateRecords(records []Record) error {
	// Check if records are continuous and valid
	if records[0].ID != 0 {
		return errors.New("no root node")
	}

	if records[0].Parent != 0 {
		return errors.New("root node has parent")
	}

	// Check for duplicate IDs, non-continuous IDs, and parent-child relationships
	for i, record := range records {
		// Check if IDs are continuous
		if record.ID != i {
			return errors.New("non-continuous node IDs")
		}

		// Check parent-child relationship
		if record.ID != 0 && record.Parent >= record.ID {
			return errors.New("parent ID must be less than child ID")
		}

		// Check for cycles
		if record.ID == record.Parent && record.ID != 0 {
			return errors.New("cycle detected")
		}
	}

	return nil
}
